nixos/modules/services/backup/syncoid.nix

```diff
@@ -138,21 +67,19 @@ in
       # Permissions snapshot and destroy are in case --no-sync-snap is not used
       default = [ "bookmark" "hold" "send" "snapshot" "destroy" ];
       description = lib.mdDoc ''
-        Permissions granted for the {option}`services.syncoid.user` user
-        for local source datasets. See
-        <https://openzfs.github.io/openzfs-docs/man/8/zfs-allow.8.html>
+        Permissions granted for the syncoid user for local source datasets.
+        See <https://openzfs.github.io/openzfs-docs/man/8/zfs-allow.8.html>
         for available permissions.
       '';
     };
 
     localTargetAllow = mkOption {
       type = types.listOf types.str;
-      default = [ "change-key" "compression" "create" "mount" "mountpoint" "receive" "rollback" ];
+      default = [ "change-key" "compression" "create" "destroy" "mount" "mountpoint" "receive" "rollback" ];
```

Comment:
There is an existing discussion of this change in https://github.com/NixOS/nixpkgs/issues/216614, and I don't think anyone has brought up this point. I've always been wary of the destroy permission as well, but I had previously thought it was mostly harmless in `localTargetAllow` because I hadn't considered the pull backup case.

The main argument for making it part of the defaults is that syncoid's default configuration (without `--no-sync-snap`) has to destroy a snapshot each time it runs.

Reply:
I was not aware of https://github.com/NixOS/nixpkgs/issues/216614. As described in 2035c7fcbc5b2d9ea1ffbca5246f498d9adc1ee0 (body of the commit message), adding `destroy` to `localTargetAllow` is required to reset broken receive states (`zfs receive -A`), which syncoid does when it fails to resume a receive state, because the snapshot has been destroyed on the source (which happened here to me on an `_hourly` sanoid snapshot which had been destroyed by a sanoid policy between the first run of syncoid which was interrupted for whatever reason, and the new one).

To confirm this, I just reproduced it like so:
```console
# Creating a source dataset, @file will be destroyed
root@losurdo$ zfs create das1/julm/tank
root@losurdo$ zfs snapshot das1/julm/tank@init
root@losurdo$ dd if=/dev/urandom of=/mnt/das1/julm/tank/file.rand status=progress count=10 bs=10MiB
root@losurdo$ zfs snapshot das1/julm/tank@file

# Creating a target dataset, `destroy` is not yet allowed to `julm`
root@oignon$ zfs create oignon/backup
root@oignon$ zfs allow julm change-key,compression,create,mount,mountpoint,receive,rollback oignon/backup
root@oignon$ chmod a+rwx /backup

# Testing the receiving
julm@oignon$ syncoid --no-privilege-elevation --no-sync-snap losurdo.wg:das1/julm/tank oignon/backup/tank
NEWEST SNAPSHOT: file
Sending incremental das1/julm/tank@init ... file (~ 100.2 MB):
^C

# Testing the resuming, and aborting again
julm@oignon$ syncoid --no-privilege-elevation --no-sync-snap losurdo.wg:das1/julm/tank oignon/backup/tank
Resuming interrupted zfs send/receive from das1/julm/tank to oignon/backup/tank (~ 100.1 MB remaining):
^C

# Destroying the source snapshot used by the receiving state
root@losurdo$ zfs destroy das1/julm/tank@file

# Trying to resume again: sending fails to `destroy` the receiving state
julm@oignon$ syncoid --debug --no-privilege-elevation --no-sync-snap losurdo.wg:das1/julm/tank oignon/backup/tank
DEBUG: SSHCMD: ssh     
DEBUG: checking availability of lzop on source...
DEBUG: checking availability of lzop on target...
DEBUG: checking availability of lzop on local machine...
DEBUG: checking availability of mbuffer on source...
DEBUG: checking availability of mbuffer on target...
DEBUG: checking availability of pv on local machine...
DEBUG: checking availability of zfs resume feature on source...
DEBUG: checking availability of zfs resume feature on target...
DEBUG: syncing source das1/julm/tank to target oignon/backup/tank.
DEBUG: getting current value of syncoid:sync on das1/julm/tank...
ssh      -S /tmp/syncoid-losurdo.wg-1696281307-7975 losurdo.wg  zfs get -H syncoid:sync ''"'"'das1/julm/tank'"'"''
DEBUG: checking to see if oignon/backup/tank on  is already in zfs receive using  ps -Ao args= ...
DEBUG: checking to see if target filesystem exists using "  zfs get -H name 'oignon/backup/tank' 2>&1 |"...
DEBUG: getting current value of receive_resume_token on oignon/backup/tank...
  zfs get -H receive_resume_token 'oignon/backup/tank'
DEBUG: got receive resume token: 1-de168d8de-e8-789c636064000310a501c49c50360710a715e5e7a69766a63040816a07d7572d29cdf50a40363b92bafca4acd4e412060616a83a0cf9b4b4e2d412063800c9b321c9275596a41603e90fbc2c58f597e4435c61794cd6e947d5ad670648f29c60f9bcc4dc54068694c46243fdacd29c5cfd92c4bc6c87b4cc9c54b89d00c9b12039: 
DEBUG: getting estimated transfer size from source -S /tmp/syncoid-losurdo.wg-1696281307-7975 losurdo.wg using "ssh      -S /tmp/syncoid-losurdo.wg-1696281307-7975 losurdo.wg  zfs send  -nvP -t 1-de168d8de-e8-789c636064000310a501c49c50360710a715e5e7a69766a63040816a07d7572d29cdf50a40363b92bafca4acd4e412060616a83a0cf9b4b4e2d412063800c9b321c9275596a41603e90fbc2c58f597e4435c61794cd6e947d5ad670648f29c60f9bcc4dc54068694c46243fdacd29c5cfd92c4bc6c87b4cc9c54b89d00c9b12039 2>&1 |"...
DEBUG: sendsize = 
Resuming interrupted zfs send/receive from das1/julm/tank to oignon/backup/tank (~ UNKNOWN remaining):
DEBUG: ssh      -S /tmp/syncoid-losurdo.wg-1696281307-7975 losurdo.wg ' zfs send  -t 1-de168d8de-e8-789c636064000310a501c49c50360710a715e5e7a69766a63040816a07d7572d29cdf50a40363b92bafca4acd4e412060616a83a0cf9b4b4e2d412063800c9b321c9275596a41603e90fbc2c58f597e4435c61794cd6e947d5ad670648f29c60f9bcc4dc54068694c46243fdacd29c5cfd92c4bc6c87b4cc9c54b89d00c9b12039 | lzop  | mbuffer  -q -s 128k -m 16M' | mbuffer  -q -s 128k -m 16M | lzop -dfc | pv -p -t -e -r -b -s 0 |  zfs receive  -s -F 'oignon/backup/tank' 2>&1
cannot resume send: 'das1/julm/tank@file' used in the initial send no longer exists
cannot receive: failed to read from stream
WARN: resetting partially receive state because the snapshot source no longer exists
DEBUG: reset partial receive state of oignon/backup/tank...
  zfs receive -A 'oignon/backup/tank'
cannot destroy 'oignon/backup/tank/%recv': permission denied
CRITICAL ERROR:   zfs receive -A 'oignon/backup/tank' failed: 256 at /nix/store/8qlzmagqra5vrkm80hm0lf16wg1095nx-sanoid-2.2.0/bin/.syncoid-wrapped line 2177.

# Allowing `julm` to `destroy`
root@oignon$ sudo zfs allow julm destroy oignon/backup/tank
# Now sending works again
julm@oignon$ syncoid --debug --no-privilege-elevation --no-sync-snap losurdo.wg:das1/julm/tank
DEBUG: SSHCMD: ssh     
DEBUG: checking availability of lzop on source...
DEBUG: checking availability of lzop on target...
DEBUG: checking availability of lzop on local machine...
DEBUG: checking availability of mbuffer on source...
DEBUG: checking availability of mbuffer on target...
DEBUG: checking availability of pv on local machine...
DEBUG: checking availability of zfs resume feature on source...
DEBUG: checking availability of zfs resume feature on target...
DEBUG: syncing source das1/julm/tank to target oignon/backup/tank.
DEBUG: getting current value of syncoid:sync on das1/julm/tank...
ssh      -S /tmp/syncoid-losurdo.wg-1696281928-2345 losurdo.wg  zfs get -H syncoid:sync ''"'"'das1/julm/tank'"'"''
DEBUG: checking to see if oignon/backup/tank on  is already in zfs receive using  ps -Ao args= ...
DEBUG: checking to see if target filesystem exists using "  zfs get -H name 'oignon/backup/tank' 2>&1 |"...
DEBUG: getting current value of receive_resume_token on oignon/backup/tank...
  zfs get -H receive_resume_token 'oignon/backup/tank'
DEBUG: got receive resume token: 1-de168d8de-e8-789c636064000310a501c49c50360710a715e5e7a69766a63040816a07d7572d29cdf50a40363b92bafca4acd4e412060616a83a0cf9b4b4e2d412063800c9b321c9275596a41603e90fbc2c58f597e4435c61794cd6e947d5ad670648f29c60f9bcc4dc54068694c46243fdacd29c5cfd92c4bc6c87b4cc9c54b89d00c9b12039: 
DEBUG: getting estimated transfer size from source -S /tmp/syncoid-losurdo.wg-1696281928-2345 losurdo.wg using "ssh      -S /tmp/syncoid-losurdo.wg-1696281928-2345 losurdo.wg  zfs send  -nvP -t 1-de168d8de-e8-789c636064000310a501c49c50360710a715e5e7a69766a63040816a07d7572d29cdf50a40363b92bafca4acd4e412060616a83a0cf9b4b4e2d412063800c9b321c9275596a41603e90fbc2c58f597e4435c61794cd6e947d5ad670648f29c60f9bcc4dc54068694c46243fdacd29c5cfd92c4bc6c87b4cc9c54b89d00c9b12039 2>&1 |"...

DEBUG: sendsize = 
Resuming interrupted zfs send/receive from das1/julm/tank to oignon/backup/tank (~ UNKNOWN remaining):
DEBUG: ssh      -S /tmp/syncoid-losurdo.wg-1696281928-2345 losurdo.wg ' zfs send  -t 1-de168d8de-e8-789c636064000310a501c49c50360710a715e5e7a69766a63040816a07d7572d29cdf50a40363b92bafca4acd4e412060616a83a0cf9b4b4e2d412063800c9b321c9275596a41603e90fbc2c58f597e4435c61794cd6e947d5ad670648f29c60f9bcc4dc54068694c46243fdacd29c5cfd92c4bc6c87b4cc9c54b89d00c9b12039 | lzop  | mbuffer  -q -s 128k -m 16M' | mbuffer  -q -s 128k -m 16M | lzop -dfc | pv -p -t -e -r -b -s 0 |  zfs receive  -s -F 'oignon/backup/tank' 2>&1
cannot resume send: 'das1/julm/tank@file' used in the initial send no longer exists
cannot receive: failed to read from stream
WARN: resetting partially receive state because the snapshot source no longer exists
DEBUG: reset partial receive state of oignon/backup/tank...
  zfs receive -A 'oignon/backup/tank'
DEBUG: syncing source das1/julm/tank to target oignon/backup/tank.
DEBUG: getting current value of syncoid:sync on das1/julm/tank...
ssh      -S /tmp/syncoid-losurdo.wg-1696281928-2345 losurdo.wg  zfs get -H syncoid:sync ''"'"'das1/julm/tank'"'"''
DEBUG: checking to see if oignon/backup/tank on  is already in zfs receive using  ps -Ao args= ...
DEBUG: checking to see if target filesystem exists using "  zfs get -H name 'oignon/backup/tank' 2>&1 |"...
DEBUG: getting current value of receive_resume_token on oignon/backup/tank...
  zfs get -H receive_resume_token 'oignon/backup/tank'
DEBUG: no receive token found 
DEBUG: getting list of snapshots on das1/julm/tank using ssh      -S /tmp/syncoid-losurdo.wg-1696281928-2345 losurdo.wg  zfs get -Hpd 1 -t snapshot guid,creation ''"'"'das1/julm/tank'"'"'' |...
DEBUG: getting list of snapshots on oignon/backup/tank using   zfs get -Hpd 1 -t snapshot guid,creation 'oignon/backup/tank' |...
NEWEST SNAPSHOT: init
DEBUG: getting current value of -p used on oignon/backup/tank...
  zfs get -H -p used 'oignon/backup/tank'
DEBUG: checking to see if oignon/backup/tank on  is already in zfs receive using  ps -Ao args= ...
INFO: no snapshots on source newer than init on target. Nothing to do, not syncing.
```

[zfs-allow(8)](https://openzfs.github.io/openzfs-docs/man/master/8/zfs-allow.8.html) doesn't seem to give a finer permission than `destroy`.